src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,916 @@
+    w2 = rho_p * v1
+    w3 = rho_p * v2
+    w4 = rho_p * v3
+    w5 = -rho_p
+    rho_p_plus += rho_p
+
+    set_component!(entropy, k, w1, w2, w3, w4, w5, equations)
+  end
+
+  # Additional non-conservative variables
+  entropy[1] = rho_p_plus * B1
+  entropy[2] = rho_p_plus * B2
+  entropy[3] = rho_p_plus * B3
+  
+  return SVector(entropy)
+end
+
+
+"""
+Convert primitive to conservative variables
+"""
+@inline function prim2cons(prim, equations::IdealMhdMultiIonEquations2D)
+  @unpack gammas = equations
+  B1, B2, B3, _ = prim
+
+  cons = zero(MVector{nvariables(equations), eltype(prim)})
+  cons[1] = B1
+  cons[2] = B2
+  cons[3] = B3
+  for k in eachcomponent(equations)
+    rho, v1, v2, v3, p = get_component(k, prim, equations)
+    rho_v1 = rho * v1
+    rho_v2 = rho * v2
+    rho_v3 = rho * v3
+
+    rho_e = p/(gammas[k] - 1.0) + 0.5 * (rho_v1 * v1 + rho_v2 * v2 + rho_v3 * v3) +
+                                  0.5 * (B1^2 + B2^2 + B3^2)
+    
+    set_component!(cons, k, rho, rho_v1, rho_v2, rho_v3, rho_e, equations)
+  end
+
+  return SVector(cons)
+end
+
+"""
+Compute the fastest wave speed for ideal MHD equations: c_f, the fast magnetoacoustic eigenvalue
+  !!! ATTENTION: This routine is provisional.. Change once the fastest wave speed is known!!
+"""
+@inline function calc_fast_wavespeed(cons, orientation::Integer, equations::IdealMhdMultiIonEquations2D)
+  B1, B2, B3, _ = cons
+
+  c_f = zero(real(equations))
+  for k in eachcomponent(equations)
+    rho, rho_v1, rho_v2, rho_v3, rho_e = get_component(k, cons, equations)
+    
+    v1 = rho_v1 / rho
+    v2 = rho_v2 / rho
+    v3 = rho_v3 / rho
+    v_mag = sqrt(v1^2 + v2^2 + v3^2)
+    gamma = equations.gammas[k]
+    p = (gamma - 1)*(rho_e - 0.5*rho*v_mag^2 - 0.5*(B1^2 + B2^2 + B3^2))
+    a_square = gamma * p / rho
+    sqrt_rho = sqrt(rho)
+
+    b1 = B1 / sqrt_rho
+    b2 = B2 / sqrt_rho
+    b3 = B3 / sqrt_rho
+    b_square = b1^2 + b2^2 + b3^2
+
+    if orientation == 1
+      c_f = max(c_f, sqrt(0.5*(a_square + b_square) + 0.5*sqrt((a_square + b_square)^2 - 4.0*a_square*b1^2)))
+    else #if orientation == 2
+      c_f = max(c_f, sqrt(0.5*(a_square + b_square) + 0.5*sqrt((a_square + b_square)^2 - 4.0*a_square*b2^2)))
+    end
+  end
+
+  return c_f
+end
+
+"""
+Routine to compute the Charge-averaged velocities:
+* v*_plus: Charge-averaged velocity
+* vk*_plus: Contribution of each species to the charge-averaged velocity
+"""
+@inline function charge_averaged_velocities(u, equations::IdealMhdMultiIonEquations2D)
+
+  total_electron_charge = zero(real(equations))
+  
+  vk1_plus = zero(MVector{ncomponents(equations), eltype(u)})
+  vk2_plus = zero(MVector{ncomponents(equations), eltype(u)})
+  vk3_plus = zero(MVector{ncomponents(equations), eltype(u)})
+
+  for k in eachcomponent(equations)
+    rho_k = u[(k-1)*5+4]
+    rho_v1_k = u[(k-1)*5+5]
+    rho_v2_k = u[(k-1)*5+6]
+    rho_v3_k = u[(k-1)*5+7]
```

Comment:
```suggestion
    rho_k, rho_v1_k, rho_v2_k, rho_v3_k, _ = get_component(k, u, equations)
```
Maybe for clarity (here and elsewhere)? Also, this would allow you to be agnostic of the internal variable order

Reply:
Also, elsewhere you use
```julia
rho, rho_v1, rho_v2, rho_v3, _ = get_component(k, u, equations)
```
e.g. in 
https://github.com/trixi-framework/Trixi.jl/pull/1427/files#diff-79474ff105011f9b8efd24e29d8d6f299c45b4ed32f19e0c6aff6ca2b7433d60R144

i.e., without the `_k` suffix. I suggest consistency 😎

Reply:
Does the `get_component` function return pointers or copies of the entries? If it returns copies, then
```
rho_k, rho_v1_k, rho_v2_k, rho_v3_k, _ = get_component(k, u, equations)
```
could affect performance, as it would copy an extra variable that is not needed

Reply:
I implemented it in e6fc017 and didn't see any change in the allocations

Reply:
There will not be any allocations either way, since everything happening with `SVector`s or tuples is on the stack. There might be extra copies of memory, but I'd assume that the compiler is smart enough to avoid copies of bits on the stack if it can reason about the data not being accessed anymore.

Either way, I think it is something that should only be optimized by hand if it shows up in a performance analysis step.